### DISCUSSION TITLE
Asynchronous or Reactive support for JPA

### DISCUSSION THREAD

**Discussion by trajano** [jakartaee/persistence#445] [Ideas]
I didn't want to create this as an issue yet.  I was thinking one of the limitations we have in Spring Data JPA was we cannot use it in a Project Reactor cleanly.

I don't even know how one would do this because JDBC isn't compatible with reactive which is why I see R2DBC.  Not sure if JDBC can be async either.

**Comment by gavinking** [comment]
Have you tried Hibernate Reactive?

**Reply by trajano** [reply]
Nope never heard of it when I was exploring Reactive development with databases a couple of years ago.  I am not sure if is part of the Spring Boot ecosystem yet.

I don't think it is though because of this line

> Hibernate Reactive is intended for use in a reactive programming environment like [Vert.x](https://vertx.io/) or [Quarkus](https://quarkus.io/), where interaction with the database should occur in a non-blocking fashion.

Since I think Spring Boot uses Project Reactor

**Reply by hantsy** [reply]
It is easy to use hibernate reactive in Spring applications, Spring WebFlux use Reactor by default, but supports Rxjava and Smallrye Munity seamlessly.

Check my blog entry about Hibernate Reactive and Spring integration, https://itnext.io/integrating-hibernate-reactive-with-spring-5427440607fe

**Reply by hantsy** [reply]
>  I am not sure if is part of the Spring Boot ecosystem yet.

Created issues on requesting integration with Spring/Spring Boot for years, but the Spring guys rejected it.

**Comment by lprimak** [comment]
With release of JDK 21 and virtual threads, standardizing asynchronous programming model in specs such as JPA is no longer necessary.

**Reply by hantsy** [reply]
Virtual thread is another topic, Reactive Streams provide more features other than async, eg. Pub/sub pattern, Message driven, back pressure, etc.

**Reply by lprimak** [reply]
However, all of those features come with a great cost, i.e. loss of imperative programming model.
Most, if not all of those features are available now with CDI, MDB (future MDB) etc?
Worth it? Only time will tell. 
If I was a betting man, I would say no.
